### ISSUE TITLE
[Cell Editor] Data detection / Qt edit gadget strip [CR's] from BLOB data.

### ISSUE THREAD

**Issue by sky5walk** [sqlitebrowser/sqlitebrowser#1793]
Details for the issue
--------------------
Bug taken from [Closed Issue](https://github.com/sqlitebrowser/sqlitebrowser/issues/1772#issuecomment-470671049)
#### What did you do?
Create a BLOB column.
Populate with any multiline text.
View entry in Cell Editor and observe with the available options:
Text
Binary
Image
XML
JSON

#### What did you expect to see?
When selecting Binary, I expected to see the EXACT hex information I provided.
And the exact Byte count I provided.

#### What did you see instead?
The hex information shows my [CR] characters are stripped from the [CR+LF] line endings.
Only Hex 0A remains.
The DB4S Character(s) count is NOT accurate since the [CR]'s are missing.

Useful extra information
-------------------------
The Qt editor/DB4S code does reapply the [CR+LF] characters on export or cut and paste clipboard?
All very confusing.
The prior Issue mentioned dropping the Qt editor gadget and just use the Scintilla gadget for all modes.
#### What operating system are you using?

- [x] Windows: ( _version:_10 Pro___ )
- [ ] Linux: ( _distro:_ ___ )
- [ ] Mac OS: ( _version:_ ___ )
- [ ] Other: ___

#### What is your DB4S version?

- [ ] 3.11.1
- [ ] 3.11.0
- [ ] 3.10.1
- [x] Other: ___3.11.99 v190304

#### Did you also

- [x] Try out the latest nightly version: https://github.com/sqlitebrowser/sqlitebrowser#nightly-builds
- [x] Search for an existing similar issue: https://github.com/sqlitebrowser/sqlitebrowser/issues?utf8=%E2%9C%93&q=is%3Aissue%20

**Comment by mgrojo** [comment]
Thanks @sky5walk. Have you confirmed that the Scintilla widget can load a Windows text file (a JSON or XML file with CR+LF endings) and behaves correctly in Windows, both in terms of displaying the lines and CR not being stripped?

**Comment by sky5walk** [comment]
Yes, but with a simple test:
1. Navigate to a null BLOB cell.
2. SET cell editor mode to JSON
3. Paste/Import json file contents with [CRLF] line endings.
4. Observe Character(s) calculated by DB4S.
5. Repeat 1-4 with other cell editor modes.

NOTE: Once the [CR] characters are stripped, the character count is invalidated.
           Pasting with Binary cell mode corrupts the data?
           I don't understand Binary?

INPUT -> JSON w/[CRLF] from Clipboard
--
CellMode | Character(s)
-- | --
TEXT | FAIL <-- [CR]'s stripped
JSON | PASS
XML | PASS <-- Auto-detection reassigns mode to JSON.
BINARY | FAIL <-- Why is input data altered?

INPUT -> JSON w/[CRLF] from imported file.
--
CellMode | Character(s)
-- | --
TEXT | FAIL <-- [CR]'s stripped
JSON | PASS
XML | PASS <-- Auto-detection reassigns mode to JSON.
BINARY | PASS <-- Auto-detection reassigns mode to JSON.

Conclusion:
--
Auto-detection should be disabled when user selects BINARY mode.

**Comment by sky5walk** [comment]
Yay 😄, the cell editor Scintilla gadget does not strip [CR]'s and I now have identical Character counts.
BUG(s): 
If I hit the auto-detect wheel, the font is changed in the cell editor Scintilla gadget.
If I hit XML or any non-TEXT mode and return to TEXT, the font is changed.
It returns to my setting after closing and reopening DB4S.

Do we really need the auto-detect wheel❓

**Comment by Johntaa** [comment]
Database cell editor doesn't accept LF & CR in text , I used to insert html entity <BR> to substitute the LF.

**Comment by sky5walk** [comment]
Ooops, [LF]'s are converted to [CRLF]'s now. 😮 
Still so much auto-detection??
Even if I paste into BLOB with Binary mode.

**Comment by mgrojo** [comment]
>Ooops, [LF]'s are converted to [CRLF]'s now. :open_mouth:
Still so much auto-detection??

If the only problem is when pasting, I think that using this Scintilla option will fix it:
https://www.scintilla.org/ScintillaDoc.html#SCI_GETPASTECONVERTENDINGS

The drawback is that pasting data from sources of different modes will convert the text in a mixture. This will probably be unusual when working in only one OS with the same DB.

> If I hit the auto-detect wheel, the font is changed in the cell editor Scintilla gadget.
If I hit XML or any non-TEXT mode and return to TEXT, the font is changed.
It returns to my setting after closing and reopening DB4S.

I can't reproduce this. The text mode is supposed to use the same font as the Data Browser while the other modes use the same font as the SQL editor (a monotype font, more adequate for a data definition or programming language). This was already so, but now it is trickier because the widget is the same and the font has to switch. I had problems with this, but now is working good for me, so once again a platform issue?

**Comment by sky5walk** [comment]
Yes, it seems to respect the content with File Import.

Only TEXT mode changes the font.
The other modes keep my original font.

**Comment by mgrojo** [comment]
You're right, the font used in the text mode is not the one configured for the data browser, as before, I think it's the font used by default by QScintilla. It is ignoring the setting. I'll have to make some testing or ask in the QScintilla mailing list.

**Comment by mgrojo** [comment]
It should be fixed in tomorrow's nightly build.

**Comment by sky5walk** [comment]
Nightly v190311 fixes the font change event. 👍 
If @Johntaa is ok with the LF issue, (I do not have his use model), I would close this issue.

**Comment by justinclift** [comment]
@Johntaa Would you have time to check the nightly builds, and see if the problem is fixed from your point of view? :smile:

If you do, the nightly builds are here:

&nbsp; &nbsp; https://nightlies.sqlitebrowser.org/latest/

**Comment by Johntaa** [comment]
@justinclift  , Sorry, I tried many times but the nightly builds cannot show the text field any more in "Edit Data Cell" window, only if the language of the text is English, but not if text is in Arabic. This is  a new issue that was not exist in the earlier releases. you can try the attached file, check the text field.
[sample.zip](https://github.com/sqlitebrowser/sqlitebrowser/files/3000892/sample.zip)

**Comment by justinclift** [comment]
Thanks @Johntaa, that's really useful to know.  We'll need to fix that. :smile:

**Comment by mgrojo** [comment]
Bad news. We knew there were problems with mixing RTL and LTR characters (#884), but this is worst even though the text is all Arabic. Interestingly, the text is displayed in the JSON mode (although left aligned). [I've written in the QScintilla mailing list](https://www.riverbankcomputing.com/pipermail/qscintilla/2019-March/001382.html).

**Comment by mgrojo** [comment]
I received answer from the QScintilla author saying that there are no short term plans to use the Scintilla version that fixes problems with right-to-left languages. We are stuck between having problems and lack of features in the Qt editor and this problem in the Scintilla editor.

Maybe we could try to backport the solution of https://sourceforge.net/p/scintilla/bugs/1590/ to our copy of QScintilla? Someone knows how to locate that changes in the mercurial history of Scintilla I didn't find a way in sourceforge.

**Comment by justinclift** [comment]
Oh damn.  That's "between a rock and a hard place" situation. :face_with_head_bandage: 

> Maybe we could try to backport the solution of https://sourceforge.net/p/scintilla/bugs/1590/ to our copy of QScintilla?

That's probably the only feasible approach.  And for longer term sanity, probably trying to contribute the backported changes to the main QScintilla project too, so we don't diverge over time.

Reading the response to your email, it's *probably* a case of them not having short term plans to do so, because it's likely to be a non-trivial amount of effort.  But that's just me guessing.

It's probably wise to ask whether they'd accept such a contribution, if someone (eg us) does the porting work.  Assuming reasonable code quality of course. :smile:

**Comment by mgrojo** [comment]
I consider this problem with RTL scripts as a release blocker for 3.12. The option to update QScintilla to a newer Scintilla version seems complex, and reverting to Qt widget would be a step forward for users of LTR scripts.

I think the more practical solution would be to recover the Qt widget only for RTL users. There are several ways to do it:
* Detect RTL scripts and fall back to Qt widget.
* Let users select the widget for text in preferences.
* Let users select the widget for text in the drop-down Mode list.

Since the only reason to fall back to Qt is that RTL is not supported, I would go for detecting RTL. But the switch to the proper widget can be problematic. If the user starts to type and the entered string contains RTL characters, the switch should be made on the fly and that would be difficult to implement right. So maybe the best option is to let the user select the widget, but if feels so wrong to ask the user for that technical choice... Any opinions on this?

**Comment by scottfurry** [comment]
This should be apart of Locale detection. IIRC, detection of RTL encoding is apart of the user's choice for localization for their operating system.

**Comment by scottfurry** [comment]
Google search: https://doc.qt.io/qt-5/qguiapplication.html#layoutDirection-prop

**Comment by mgrojo** [comment]
You're right. I was thinking in using either the [QChar::script()](https://doc.qt.io/qt-5/qchar.html#script) function (I think only Arabic, Syriac and Hebrew in that list are RTL scripts) or the [QChar::direction()](https://doc.qt.io/qt-5/qchar.html#direction) function (R, AL, RLE, RLO, RLI) are RTL characters, according to [Unicode](http://www.unicode.org/reports/tr9/tr9-35.html#Table_Bidirectional_Character_Types).

**Comment by mgrojo** [comment]
> Google search: https://doc.qt.io/qt-5/qguiapplication.html#layoutDirection-prop

It is another option. I would be simple, but then a user with a RTL layout could be using English or another language in the DB or viceversa, an English interface in the application and a DB in Arabic or Hebrew.

**Comment by scottfurry** [comment]
A little more digging found that functionality exists in QApplication. This would be detectable at startup in the main.cpp and could be set accordingly. The layout direction would then propagate downward to all created child widgets.

**Comment by justinclift** [comment]
Sounds like it might be worth exploring. :smile:

**Comment by Johntaa** [comment]
> 
> 
> Sounds like it might be worth exploring. smile

Until you decide which option is the right one , isn't it possible to develop 3.12 as a new branch to follow up with the new updates other than Qt widget. and keep the Qt widget .
Just a Question :)

**Comment by justinclift** [comment]
Oh yeah, there's a bunch of different possibilities for doing that.  It just depends on who wants to take it up and how they want to do it. :smile:

**Comment by mgrojo** [comment]
@sky5walk Are all the problems address in 3.12 alpha?

**Comment by sky5walk** [comment]
Ok, almost done.
For existing database content, the binary view now shows the [CR][LF] characters.
But not when I paste new data into a blob field?
The [CR]'s are not deleted, but they are not shown in the binary view.

**Comment by justinclift** [comment]
As a general thought, I've been wondering if we should try building with the newest Qt release (5.14.2) too.

Apparently that has changes to better enable accessibility, which would really help with Community members needing to use screen readers and such.
